sensors: lis2dh: fix runtime and 16g range compile-time selectable range conversion

mniestroj opened the pull request:
This is an alternative implementation of PR #19983

mniestroj committed:
sensors: lis2dh: fix raw value to sensor_value conversion

According to struct sensor_value documentation, val2 should be negative
for negative result. So drop code that tries to make val2 positive.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

mniestroj committed:
sensors: lis2dh: fix raw value scaling

All values were scaled by 8 instead of 12 for 16g compile-time selected
range. This resulted with values around 6.5 m/s^2. Scaling for runtime
configurable ranges was broken for all except 2g range option due to
totally broken lis2dh_range_to_reg_val().

Fix wrong scaling for 16g compile-time and all runtime selectable ranges
by reworking code that scales raw value. While doing this, change
lis2dh->scale type from 16 to 32 bits. This allows to slightly increase
final result precision by using the fact that raw values have maximum 12
bits precision, allowing us to multiply lis2dh->scale by (1 << 4)
compared to previous implementation.

Fix bug #19872.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

avisconti commented:
This PR is addressing #19872

avisconti commented:
I'm ok withis changes, but we need to address the following:

As you probably know we are using the same driver for different devices that has same registers interface (i.e. whoami is same: 0x33). This devices currently are: LIS2DH, LIS3DH, LSM303DLHC, LIS2DH12, LSM303AGR (see drivers/sensor/lis2dh/Kconfig).

Now the issue is that, even if reg i/f is same, they have different mechanical sensors, so the scale is different. I was using the lsm303agr ones...

avisconti commented:
> I'm ok withis changes, but we need to address the following:
> 
> As you probably know we are using the same driver for different devices that has same registers interface (i.e. whoami is same: 0x33). This devices currently are: LIS2DH, LIS3DH, LSM303DLHC, LIS2DH12, LSM303AGR (see drivers/sensor/lis2dh/Kconfig).
> 
> Now the issue is that, even if reg i/f is same, they have different mechanical sensors, so the scale is different. I was using the lsm303agr ones...

As @mniestroj  noted, seems that only LSM303AGR has a different set of gains, so we need to implement this in the driver.

avisconti reviewed (changes requested):
Please address what discussed previously, i.e. addtressing the LSM303AGR set of gains.

mniestroj committed:
dts: bindings: lsm303agr: add new bindings based on lis2dh

Add new bindings for already supported (by lis2dh.c driver) lsm303agr
accelerometer. Using st,lsm303agr-accel compatible will allow in
subsequent commit to apply different sensitivity scale.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

mniestroj committed:
sensors: lis2dh: use proper scale factor for lsm303agr

LSM303AGR is a special one in terms of raw value scale among all devices
supported by lis2dh.c driver. Apply proper scale factor based on
sensitivity scale provided in LIS2DH and LSM303AGR datasheets.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

mniestroj commented:
@avisconti 
I have just rebased this MR on top of master, in order to get rid of conflicts.

@sdalu 
Could you test if this MR solves #19872 ?

avisconti commented on drivers/sensor/lis2dh/lis2dh.c:
```diff
@@ -13,6 +13,26 @@
+ * Use values for 16g low-power mode in LSM303AGR DS "Table 3. Sensor
+ * characteristics" and LIS2DH DS "Table 3. Mechanical characteristics".
+ */
+#define ACCEL_SCALE_FACTOR(x)		((x) * 18758LL / 19200LL)
```
So you decided to use a matematic expression.
Actually I preferred to have a separate set of scales for the two different sensors.
I think it is easier to find the correspondent values on the datasheets.
What do you think?

mniestroj replied:
```
#define ACCEL_SCALE(sensitivity)		\
	(SENSOR_G * (sensitivity) >> 14 / 100)

/*
 * Use values for low-power mode in DS "Mechanical (Sensor) characteristics",
 * multiplied by 100.
 */
static const u32_t lis2dh_reg_val_to_scale[] = {
#if defined(DT_INST_0_ST_LSM303AGR_ACCEL)
	ACCEL_SCALE(1563),
	ACCEL_SCALE(3126),
	ACCEL_SCALE(6252),
	ACCEL_SCALE(18758),
#else
	ACCEL_SCALE(1600),
	ACCEL_SCALE(3200),
	ACCEL_SCALE(6400),
	ACCEL_SCALE(19200),
#endif
};
```

Something like that? :)

avisconti replied:
> ```
> #define ACCEL_SCALE(sensitivity)		\
> 	(SENSOR_G * (sensitivity) >> 14 / 100)
> 
> /*
>  * Use values for low-power mode in DS "Mechanical (Sensor) characteristics",
>  * multiplied by 100.
>  */
> static const u32_t lis2dh_reg_val_to_scale[] = {
> #if defined(DT_INST_0_ST_LSM303AGR_ACCEL)
> 	ACCEL_SCALE(1563),
> 	ACCEL_SCALE(3126),
> 	ACCEL_SCALE(6252),
> 	ACCEL_SCALE(18758),
> #else
> 	ACCEL_SCALE(1600),
> 	ACCEL_SCALE(3200),
> 	ACCEL_SCALE(6400),
> 	ACCEL_SCALE(19200),
> #endif
> };
> ```
> 
> 
> Something like that? :)

Yes, exactly! :)

avisconti replied:
@mniestroj 
So, are you changing it in this way?

mniestroj replied:
@avisconti Yes, just wanted to test with my board before submission :) It is ready now.

sdalu commented:
@mniestroj just tested, got what seems correct values, it solves  #19872
